solidity/contracts/midas/WithdrawalQueue.sol

```diff
@@ -0,0 +1,289 @@
+    ) external onlyAcreBTC returns (uint256 requestId) {
+        uint256 tbtcAmount = acrebtc.convertToAssets(_shares);
+        uint256 midasShares = vault.convertToShares(tbtcAmount);
+        midasAllocator.withdraw(midasShares);
+        acrebtc.burn(_shares);
+        vaultSharesToken.approve(address(vault), midasShares);
+        uint256 midasRequestId = vault.requestRedeem(
+            midasShares,
+            address(this)
+        );
+
+        requestId = count;
+        withdrawalRequests[requestId] = WithdrawalRequest({
+            redeemer: _redeemer,
+            shares: midasShares,
+            tbtcAmount: tbtcAmount,
+            createdAt: block.timestamp,
+            completedAt: 0,
+            isCompleted: false,
+            walletPubKeyHash: _walletPubKeyHash,
+            midasRequestId: midasRequestId
+        });
+
+        emit WithdrawalRequestCreated(
+            requestId,
+            _redeemer,
+            midasShares,
+            tbtcAmount,
+            _walletPubKeyHash,
+            midasRequestId
+        );
+        count++;
+    }
+
+    /// @notice Completes a withdrawal request.
+    /// @param _requestId ID of the withdrawal request.
+    /// @param _tbtcRedemptionData Additional data required for the tBTC redemption.
+    ///        See `redemptionData` parameter description of `Bridge.requestRedemption`
+    ///        function.
+    function completeWithdrawalRequest(
+        uint256 _requestId,
+        bytes calldata _tbtcRedemptionData
+    ) external onlyMaintainer {
+        // TBTC Token contract owner resolves to the TBTCVault contract.
+        if (tbtc.owner() != tbtcVault) revert UnexpectedTbtcTokenOwner();
+        if (withdrawalRequests[_requestId].isCompleted)
+            revert WithdrawalRequestAlreadyCompleted();
+
+        (uint256 tbtcAmount, uint256 exitFee) = _finalizeRequestandTakeExitFee(
+            _requestId
+        );
+
+        (address redeemer, bytes20 walletPubKeyHash, , , , ) = abi.decode(
+            _tbtcRedemptionData,
+            (address, bytes20, bytes32, uint32, uint64, bytes)
+        );
+        if (
+            redeemer != withdrawalRequests[_requestId].redeemer ||
+            walletPubKeyHash != withdrawalRequests[_requestId].walletPubKeyHash
+        )
+            revert InvalidRedemptionData(
+                redeemer,
+                withdrawalRequests[_requestId].redeemer,
+                walletPubKeyHash,
+                withdrawalRequests[_requestId].walletPubKeyHash
+            );
+
+        if (
+            !tbtc.approveAndCall(
+                tbtcVault,
+                tbtcAmount - exitFee,
+                _tbtcRedemptionData
+            )
+        ) {
+            revert ApproveAndCallFailed();
+        }
+
+        emit WithdrawalRequestCompleted(
+            _requestId,
+            tbtcAmount - exitFee,
+            exitFee
+        );
+    }
+
+    function updateTbtcVault(address _tbtcVault) external onlyMaintainer {
```

Comment:
This function and `tbtcVault` storage slot could be removed if we move tBTC Bridge integration logic to the BitcoinRedeemer contract, the way I proposed in another comment.